delphi-checks/src/main/java/au/com/integradev/delphi/checks/ConsecutiveVisibilitySectionCheck.java

```diff
@@ -0,0 +1,158 @@
+    return "visibility";
+  }
+
+  @Override
+  protected boolean areViolatingSections(DelphiNode first, DelphiNode second) {
+    if (!(first instanceof VisibilitySectionNode && second instanceof VisibilitySectionNode)) {
+      return false;
+    }
+
+    VisibilitySectionNode firstSection = (VisibilitySectionNode) first;
+    VisibilitySectionNode secondSection = (VisibilitySectionNode) second;
+
+    // Sections with different visibilities are OK
+    if (firstSection.getVisibility() != secondSection.getVisibility()) {
+      return false;
+    }
+
+    Optional<DelphiNode> firstItem = getOnlyItem(firstSection);
+    Optional<DelphiNode> secondItem = getOnlyItem(secondSection);
+
+    if (firstItem.isPresent() && secondItem.isPresent()) {
+      return !areValidConsecutiveItems(firstItem.get(), secondItem.get());
+    } else if (firstItem.isPresent()) {
+      return !isValidConsecutiveItemWithMultiItemSection(firstItem.get());
+    } else if (secondItem.isPresent()) {
+      return !isValidConsecutiveItemWithMultiItemSection(secondItem.get());
+    } else {
+      return true;
+    }
+  }
+
+  private boolean isValidConsecutiveItemWithMultiItemSection(DelphiNode item) {
+    return item instanceof TypeSectionNode
+        || item instanceof ConstSectionNode
+        || (item instanceof FieldSectionNode && ((FieldSectionNode) item).isClassFieldSection());
+  }
+
+  private boolean areValidConsecutiveItems(DelphiNode firstItem, DelphiNode secondItem) {
+    return
+    // Type -> Not type
+    (firstItem instanceof TypeSectionNode && !(secondItem instanceof TypeSectionNode))
+        // Not type -> Type
+        || (!(firstItem instanceof TypeSectionNode) && secondItem instanceof TypeSectionNode)
+        // Const -> Not const
+        || (firstItem instanceof ConstSectionNode && !(secondItem instanceof ConstSectionNode))
+        // Not const -> const
+        || (!(firstItem instanceof ConstSectionNode) && secondItem instanceof ConstSectionNode)
+        // Class var -> Not class var (or vice versa)
+        || (firstItem instanceof FieldSectionNode
+            && secondItem instanceof FieldSectionNode
+            && ((FieldSectionNode) firstItem).isClassFieldSection()
+                != ((FieldSectionNode) secondItem).isClassFieldSection());
+  }
+
+  private Optional<DelphiNode> getOnlyItem(VisibilitySectionNode visibilitySection) {
+    List<DelphiNode> items =
+        visibilitySection.getChildren().stream()
+            .filter(x -> !(x instanceof VisibilityNode))
+            .collect(Collectors.toList());
+
+    if (items.size() == 1) {
+      return Optional.of(items.get(0));
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  @Override
+  protected FilePosition getViolatingFilePosition(DelphiNode node) {
+    if (node instanceof VisibilitySectionNode) {
+      DelphiNode firstChild = node.getChild(0);
+      if (firstChild instanceof VisibilityNode) {
+        return FilePosition.from(firstChild);
```

Comment:
We can safely do these casts and also safely assume that the `VisibilitySectionNode` has a visibility node, because:
* We control what gets passed to this method in `areViolatingSections`
* We select the 2nd section as the violating node, and an `IMPLICIT_PUBLISHED` visibility section can only ever be the first one.